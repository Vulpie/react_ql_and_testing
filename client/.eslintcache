[{"/home/vulpie/Desktop/projects/ql/client/src/index.js":"1","/home/vulpie/Desktop/projects/ql/client/src/App.js":"2","/home/vulpie/Desktop/projects/ql/client/src/Components/Home.js":"3","/home/vulpie/Desktop/projects/ql/client/src/Components/Lists/BookList.js":"4","/home/vulpie/Desktop/projects/ql/client/src/Components/Lists/AuthorList.js":"5","/home/vulpie/Desktop/projects/ql/client/src/Components/Book/Book.js":"6","/home/vulpie/Desktop/projects/ql/client/src/Components/Author/Author.js":"7","/home/vulpie/Desktop/projects/ql/client/src/Components/Author/AddAuthor.js":"8","/home/vulpie/Desktop/projects/ql/client/src/Components/Partials/LinkButton.js":"9","/home/vulpie/Desktop/projects/ql/client/src/Components/Partials/HomeButton.js":"10","/home/vulpie/Desktop/projects/ql/client/src/Components/Book/AddBook.js":"11"},{"size":209,"mtime":1610839960276,"results":"12","hashOfConfig":"13"},{"size":1107,"mtime":1611438731037,"results":"14","hashOfConfig":"13"},{"size":315,"mtime":1610894367685,"results":"15","hashOfConfig":"13"},{"size":644,"mtime":1611421730355,"results":"16","hashOfConfig":"13"},{"size":613,"mtime":1611402966557,"results":"17","hashOfConfig":"13"},{"size":1147,"mtime":1611430961232,"results":"18","hashOfConfig":"13"},{"size":243,"mtime":1610889405967,"results":"19","hashOfConfig":"13"},{"size":1039,"mtime":1610878985504,"results":"20","hashOfConfig":"13"},{"size":823,"mtime":1611207371678,"results":"21","hashOfConfig":"13"},{"size":303,"mtime":1611354760355,"results":"22","hashOfConfig":"13"},{"size":1447,"mtime":1611438891194,"results":"23","hashOfConfig":"13"},{"filePath":"24","messages":"25","errorCount":0,"warningCount":0,"fixableErrorCount":0,"fixableWarningCount":0,"usedDeprecatedRules":"26"},"1k8yqt7",{"filePath":"27","messages":"28","errorCount":0,"warningCount":0,"fixableErrorCount":0,"fixableWarningCount":0},{"filePath":"29","messages":"30","errorCount":0,"warningCount":0,"fixableErrorCount":0,"fixableWarningCount":0,"usedDeprecatedRules":"26"},{"filePath":"31","messages":"32","errorCount":0,"warningCount":0,"fixableErrorCount":0,"fixableWarningCount":0,"usedDeprecatedRules":"26"},{"filePath":"33","messages":"34","errorCount":0,"warningCount":0,"fixableErrorCount":0,"fixableWarningCount":0,"usedDeprecatedRules":"26"},{"filePath":"35","messages":"36","errorCount":0,"warningCount":0,"fixableErrorCount":0,"fixableWarningCount":0},{"filePath":"37","messages":"38","errorCount":0,"warningCount":0,"fixableErrorCount":0,"fixableWarningCount":0,"usedDeprecatedRules":"26"},{"filePath":"39","messages":"40","errorCount":0,"warningCount":0,"fixableErrorCount":0,"fixableWarningCount":0,"usedDeprecatedRules":"26"},{"filePath":"41","messages":"42","errorCount":0,"warningCount":0,"fixableErrorCount":0,"fixableWarningCount":0,"usedDeprecatedRules":"26"},{"filePath":"43","messages":"44","errorCount":0,"warningCount":0,"fixableErrorCount":0,"fixableWarningCount":0,"usedDeprecatedRules":"26"},{"filePath":"45","messages":"46","errorCount":0,"warningCount":0,"fixableErrorCount":0,"fixableWarningCount":0},"/home/vulpie/Desktop/projects/ql/client/src/index.js",[],["47","48"],"/home/vulpie/Desktop/projects/ql/client/src/App.js",[],"/home/vulpie/Desktop/projects/ql/client/src/Components/Home.js",[],"/home/vulpie/Desktop/projects/ql/client/src/Components/Lists/BookList.js",[],"/home/vulpie/Desktop/projects/ql/client/src/Components/Lists/AuthorList.js",[],"/home/vulpie/Desktop/projects/ql/client/src/Components/Book/Book.js",[],"/home/vulpie/Desktop/projects/ql/client/src/Components/Author/Author.js",[],"/home/vulpie/Desktop/projects/ql/client/src/Components/Author/AddAuthor.js",[],"/home/vulpie/Desktop/projects/ql/client/src/Components/Partials/LinkButton.js",[],"/home/vulpie/Desktop/projects/ql/client/src/Components/Partials/HomeButton.js",[],"/home/vulpie/Desktop/projects/ql/client/src/Components/Book/AddBook.js",[],{"ruleId":"49","replacedBy":"50"},{"ruleId":"51","replacedBy":"52"},"no-native-reassign",["53"],"no-negated-in-lhs",["54"],"no-global-assign","no-unsafe-negation"]