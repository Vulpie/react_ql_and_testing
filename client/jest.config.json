{
	"name": "my-project",
	"jest": {
		"verbose": true
	},
	"setupFilesAfterEnv": ["raf/polyfill", "<rootDir>/src/tests/setupTests.js"],
	"testPathIgnorePatterns": ["<rootDir>/node_modules/"],
	"moduleNameMapper": {
		"\\.(jpg|ico|jpeg|png|gif|eot|otf|webp|svg|ttf|woff|woff2|mp4|webm|wav|mp3|m4a|aac|oga)$": "<rootDir>/src/tests/__mocks__/fileMock.js",
		"\\.(css|less)$": "<rootDir>/src/tests/__mocks__/fileMock.js"
	}
}
