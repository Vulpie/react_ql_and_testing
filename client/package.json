{
	"name": "client",
	"version": "0.1.0",
	"private": true,
	"proxy": "http://localhost:5000",
	"babel": {
		"presets": [
			"@babel/env",
			"@babel/react"
		]
	},
	"dependencies": {
		"@apollo/client": "^3.3.7",
		"@testing-library/jest-dom": "^5.11.9",
		"@testing-library/react": "^11.2.3",
		"@testing-library/user-event": "^12.6.0",
		"babel-jest": "^26.6.3",
		"enzyme": "^3.11.0",
		"enzyme-adapter-react-16": "^1.15.6",
		"enzyme-to-json": "^3.6.1",
		"graphql": "^15.4.0",
		"jest-dom": "^4.0.0",
		"react": "^17.0.1",
		"react-dom": "^17.0.1",
		"react-router-dom": "^5.2.0",
		"react-routes": "^0.2.6",
		"react-scripts": "4.0.1",
		"react-test-renderer": "^17.0.1",
		"web-vitals": "^0.2.4"
	},
	"scripts": {
		"start": "react-scripts start",
		"build": "react-scripts build",
		"test": "jest --config=jest.config.json --watch ",
		"eject": "react-scripts eject"
	},
	"eslintConfig": {
		"extends": [
			"react-app",
			"react-app/jest"
		]
	},
	"browserslist": {
		"production": [
			">0.2%",
			"not dead",
			"not op_mini all"
		],
		"development": [
			"last 1 chrome version",
			"last 1 firefox version",
			"last 1 safari version"
		]
	},
	"devDependencies": {
		"@wojtekmaj/enzyme-adapter-react-17": "^0.4.1",
		"identity-obj-proxy": "^3.0.0"
	}
}
