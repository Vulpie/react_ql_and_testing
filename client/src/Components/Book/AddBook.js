import React from 'react'
import { Redirect } from 'react-router-dom'
import { gql, useMutation } from '@apollo/client'

const ADD_BOOK = gql`
	mutation AddBook($name: String!, $genre: String!, $authorId: ID!) {
		addBook(name: $name, genre: $genre, authorId: $authorId) {
			name
			genre
			id
			author {
				id
			}
		}
	}
`

const AddBook = ({ authorId }) => {
	const [addBook, { data }] = useMutation(ADD_BOOK)

	const handleSubmit = (e) => {
		e.preventDefault()
		const book_name = e.target['book_name'].value
		let book_genre = e.target['book_genre'].value
		console.table({ book_name, book_genre })

		if (!book_name || !book_genre) {
			console.warn(`Missing name ${book_name} or genre ${book_genre}`)
			return
		}

		addBook({
			variables: {
				name: book_name,
				genre: book_genre,
				authorId: authorId,
			},
		})
	}
	if (data) {
		console.log(data)
		return (
			<Redirect
				to={{
					pathname: '/book/details',
					search: `?id=${data.addBook.id}&name=${data.addBook.name}&genre=${data.addBook.genre}&authorId=${data.addBook.author.id}`,
				}}
			/>
		)
	}

	return (
		<form onSubmit={(e) => handleSubmit(e)} className='book-form'>
			<label htmlFor='book_name'>Book title</label>
			<input type='text' name='book_name' required />

			<label htmlFor='book_genre'>Book genre</label>
			<input type='text' name='book_genre' required />

			<button type='submit'>Add book</button>
		</form>
	)
}

export default AddBook
