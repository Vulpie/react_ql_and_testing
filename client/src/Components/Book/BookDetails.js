import React from 'react'
import { gql, useQuery } from '@apollo/client'
import { useParams } from 'react-router-dom'
import HomeButton from '../Partials/HomeButton'

export const GET_BOOK = gql`
	query Book($id: ID!) {
		book(id: $id) {
			name
			genre
			author {
				name
				id
			}
		}
	}
`

const BookDetails = () => {
	// console.warn('ID:', id)
	let { id } = useParams()
	const { loading, error, data } = useQuery(GET_BOOK, {
		variables: { id },
	})

	if (loading) return <p>Loading list of authors ...</p>
	if (error) return <p>Error -.-</p>
	console.log(data)
	return (
		<div className='component'>
			<HomeButton />
			<h2 className='component__title'>
				"{data && data.book && data.book.name}"
			</h2>
			<div className='details'>
				<div className='details__item'>
					<p className='details__item_header'>Title</p>
					<p className='details__item_value'>
						"{data && data.book && data.book.name}"
					</p>
				</div>
				<div className='details__item'>
					<p className='details__item_header'>Genre</p>
					<p className='details__item_value'>
						{data && data.book && data.book.genre}
					</p>
				</div>
				<div className='details__item'>
					<p className='details__item_header'>Author</p>
					<p className='details__item_value'>
						{data && data.book && data.book.author.name}
					</p>
				</div>
			</div>
		</div>
	)
}

export default BookDetails
