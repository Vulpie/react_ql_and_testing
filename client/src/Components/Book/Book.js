import React, { useState } from 'react'
import HomeButton from '../Partials/HomeButton'
import { gql, useQuery } from '@apollo/client'
import AddBook from './AddBook'

export const GET_AUTHORS = gql`
	{
		authors {
			name
			id
		}
	}
`

const Book = () => {
	const { loading, error, data } = useQuery(GET_AUTHORS)
	const [selectedAuthor, setSelectedAuthor] = useState()

	return (
		<div>
			<h2>Book</h2>
			<HomeButton />
			{loading && <p>Loading list of authors ..</p>}
			{error && <p>Error -.-</p>}
			{data && (
				<select
					value={selectedAuthor}
					onChange={(e) => setSelectedAuthor(e.target.value)}
				>
					<option value=''></option>
					{data.authors.map(({ name, id }) => (
						<option key={'author_' + name} value={id}>
							{name}
						</option>
					))}
				</select>
			)}
			{selectedAuthor ? (
				<AddBook authorId={selectedAuthor} />
			) : (
				<p>Select the author</p>
			)}
		</div>
	)
}

export default Book
