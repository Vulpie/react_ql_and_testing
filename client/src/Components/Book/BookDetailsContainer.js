import React from 'react'
import { useParams } from 'react-router-dom'
import HomeButton from '../Partials/HomeButton'
import BookDetails from './BookDetails'

const BookDetailsContainer = () => {
	let { id } = useParams()
	console.log(id)
	return (
		<div className='component'>
			<HomeButton />
			<BookDetails id={id} />
		</div>
	)
}

export default BookDetailsContainer
