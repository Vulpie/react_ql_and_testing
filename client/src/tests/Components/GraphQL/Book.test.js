import React from 'react'
import { MockedProvider } from '@apollo/client/testing'
import { act, render } from '@testing-library/react'
import Book, { GET_AUTHORS } from '../../../Components/Book/Book'
import { BrowserRouter as Router } from 'react-router-dom'

describe('Render book component without errror when:', () => {
	test('Loading state', () => {
		const mocks = []
		const { getByText, container } = render(
			<MockedProvider mocks={mocks} addTypename={false}>
				<Router>
					<Book />
				</Router>
			</MockedProvider>
		)
		expect(getByText('Loading list of authors ..')).toBeInTheDocument()
		expect(container).toMatchSnapshot()
	})

	test('Success state', async () => {
		const mocks = [
			{
				request: { query: GET_AUTHORS },
				result: { data: { authors: [{ name: 'author1', id: 1 }] } },
			},
		]
		const { getByText, container } = render(
			<MockedProvider mocks={mocks} addTypename={false}>
				<Router>
					<Book />
				</Router>
			</MockedProvider>
		)
		await act(async () => {
			await new Promise((resolve) => setTimeout(resolve, 0))
		})

		expect(container).toMatchSnapshot()
		expect(getByText('Select the author')).toBeInTheDocument()
	})
})
