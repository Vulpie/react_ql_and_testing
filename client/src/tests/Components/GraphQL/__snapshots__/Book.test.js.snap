// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Render book component without errror when: Loading state 1`] = `
<div>
  <div>
    <h2>
      Book
    </h2>
    <a
      href="/"
    >
      <img
        alt="Return button"
        class="home-button"
        src=""
      />
    </a>
    <p>
      Loading list of authors ..
    </p>
    <p>
      Select the author
    </p>
  </div>
</div>
`;

exports[`Render book component without errror when: Success state 1`] = `
<div>
  <div>
    <h2>
      Book
    </h2>
    <a
      href="/"
    >
      <img
        alt="Return button"
        class="home-button"
        src=""
      />
    </a>
    <select>
      <option
        value=""
      />
      <option
        value="1"
      >
        author1
      </option>
    </select>
    <p>
      Select the author
    </p>
  </div>
</div>
`;
